application/common/models/EmailQueue.php

```diff
@@ -0,0 +1,219 @@
+         */
+        try {
+            $mailer = $this->getMailer();
+            $sent = $mailer->send();
+            if ($sent === 0 || $sent === false) {
+                throw new \Exception('Unable to send email', 1461011826);
+            }
+
+            /*
+             * Create event log entry if needed
+             */
+            $this->createEventLogEntry();
+
+            /*
+             * Remove entry from queue (if saved to queue) after successful send
+             */
+            $this->removeFromQueue();
+
+            /*
+             * Log success
+             */
+            $log['status'] = 'sent';
+            \Yii::warning($log, 'application');
+
+        } catch (\Exception $e) {
+            /*
+             * Send failed, attempt to queue
+             */
+            $this->attempts_count += 1;
+            $this->last_attempt = Utils::getDatetime();
+            $this->queue($log, $e);
+        }
+    }
+
+    /**
+     * Builds a mailer object from $this and returns it
+     * @return \yii\mail\MessageInterface
+     */
+    private function getMailer()
+    {
+        $mailer = \Yii::$app->mailer->compose();
+        $mailer->setFrom(\Yii::$app->params['fromEmail']);
+        $mailer->setTo($this->to_address);
+        $mailer->setSubject($this->subject);
+        $mailer->setTextBody($this->text_body);
+
+        /*
+         * Conditionally set optional fields
+         */
+        $setMethods = [
+            'setCc' => $this->cc_address,
+            'setHtmlBody' => $this->html_body,
+        ];
+        foreach ($setMethods as $method => $value) {
+            if ($value) {
+                $mailer->$method($value);
+            }
+        }
+        
+        return $mailer;
+    }
+
+    /**
+     * Creates an EventLog entry if $this->event_log_* properties are set
+     */
+    private function createEventLogEntry()
+    {
+        if ($this->event_log_topic !== null && $this->event_log_details !== null && $this->event_log_user_id !== null) {
+            EventLog::log($this->event_log_topic, $this->event_log_details, $this->event_log_user_id);
+        }
+    }
+
+    /**
+     * Queue's email for sending later, if there is an error saving to db it throws an exception
+     * @param array $log Optionally provide an array of data to be logged
+     * @param \Exception|null $previousException Optionally provide an exception if one occurred
+     *                                           resulting in queueing email
+     * @throws \Exception
+     */
+    private function queue($log = [], \Exception $previousException = null)
+    {
+        if ( ! $this->save()) {
+            /*
+             * Queue failed, log it and throw exception
+             */
+            $log['status'] = 'failed to queue';
+            $log['error'] = Json::encode($this->getFirstErrors());
+            \Yii::error($log, 'application');
+            throw new \Exception('Unable to send or queue email', 1461009236);
```

Comment:
This error message should probably just be `'Unable to queue email'`, since that's all this function tries to do. If there is an error that prevents both sending and queuing an email, the function that was trying to do both (such as `sendOrQueue`) seems like the more appropriate place for that kind of error message to originate.